dialogue-apache-hc5-client/src/main/java/com/palantir/dialogue/hc5/ApacheHttpClientChannels.java

```diff
@@ -0,0 +1,537 @@
+            String clientName,
+            PoolingHttpClientConnectionManager connectionManager) {
+        DialogueInternalWeakReducingGauge.getOrCreate(
+                taggedMetrics,
+                clientPoolSizeMetricName(clientName, "idle"),
+                pool -> pool.getTotalStats().getAvailable(),
+                LongStream::sum,
+                connectionManager);
+        DialogueInternalWeakReducingGauge.getOrCreate(
+                taggedMetrics,
+                clientPoolSizeMetricName(clientName, "leased"),
+                pool -> pool.getTotalStats().getLeased(),
+                LongStream::sum,
+                connectionManager);
+        DialogueInternalWeakReducingGauge.getOrCreate(
+                taggedMetrics,
+                clientPoolSizeMetricName(clientName, "pending"),
+                pool -> pool.getTotalStats().getPending(),
+                LongStream::sum,
+                connectionManager);
+    }
+
+    private static MetricName clientPoolSizeMetricName(String clientName, String state) {
+        return MetricName.builder()
+                .safeName("dialogue.client.pool.size")
+                .putSafeTags("client-name", clientName)
+                .putSafeTags("state", state)
+                .build();
+    }
+
+    /** Intentionally opaque wrapper type - we don't want people using the inner Apache client directly. */
+    public static final class CloseableClient implements Closeable {
+        private static final String APACHE = "apache";
+
+        private final String clientName;
+        private final CloseableHttpClient apacheClient;
+        private final PoolingHttpClientConnectionManager pool;
+        private final ResponseLeakDetector leakDetector;
+
+        @Nullable
+        private final ExecutorService executor;
+
+        private final Closer closer = Closer.create();
+
+        private CloseableClient(
+                CloseableHttpClient apacheClient,
+                @Safe String clientName,
+                PoolingHttpClientConnectionManager pool,
+                IdleConnectionEvictor connectionEvictor,
+                TaggedMetricRegistry taggedMetrics,
+                ResponseLeakDetector leakDetector,
+                @Nullable ExecutorService executor) {
+            this.clientName = clientName;
+            this.apacheClient = apacheClient;
+            this.pool = pool;
+            this.leakDetector = leakDetector;
+            this.executor = executor;
+            closer.register(() -> {
+                connectionEvictor.shutdown();
+                try {
+                    connectionEvictor.awaitTermination(Timeout.of(1L, TimeUnit.SECONDS));
+                } catch (InterruptedException interrupted) {
+                    Thread.currentThread().interrupt();
+                }
+            });
+            connectionEvictor.start();
+            closer.register(apacheClient);
+            closer.register(pool::close);
```

Comment:
this is a graceful shutdown (I think that's what we did in hc4). This method takes argument for non graceful shutdown